### PR TITLE
Make RemoteChannel iterable

### LINKED ISSUE #31177: iterate RemoteChannels
Is there any reason why we can't iterate a `RemoteChannel`.

The documentation for `Channel` shows an example using a `for x in channel` loop, whereas the analogous example for `RemoteChannel` uses a `while true` loop. Presumably this is because `iterate` is not defined for `RemoteChannel`.

`Channel` (https://docs.julialang.org/en/latest/manual/parallel-computing/#Channels-1)
```
function do_work()
           for job_id in jobs
               exec_time = rand()
               sleep(exec_time)                # simulates elapsed time doing actual work
                                               # typically performed externally.
               put!(results, (job_id, exec_time))
           end
       end;
```
`RemoteChannel` (https://docs.julialang.org/en/latest/manual/parallel-computing/#Channels-and-RemoteChannels-1)

```
@everywhere function do_work(jobs, results) # define work function everywhere
           while true
               job_id = take!(jobs)
               exec_time = rand()
               sleep(exec_time) # simulates elapsed time doing actual work
               put!(results, (job_id, exec_time, myid()))
           end
       end
```

Naively copying the iterate code from `Channel` to `RemoteChannel` appears to work for a simple case:
```
using Distributed

function Base.iterate(c::RemoteChannel, state=nothing)
    try
        return (take!(c), nothing)
    catch e
        if isa(e, InvalidStateException) && e.state==:closed
            return nothing
        else
            rethrow(e)
        end
    end
end

Base.IteratorSize(::Type{<:RemoteChannel}) = Base.SizeUnknown()

c = Channel{Int}(10)
foreach(i->put!(c, i), 1:3)
close(c)
data = [i for i in c]

rc = RemoteChannel(() -> Channel{Int}(10))
foreach(i->put!(rc, i), 1:3)
close(rc)
data = [i for i in rc]
```

### PR DESCRIPTION
**Pull request by tkf**
Being an iterator is one of the nice features of Channel.  Can RemoteChannel have it?

close #31177

### CODE REVIEW

**Comment by jonas-schulze** [comment]
bump 🙂

**Comment by StefanKarpinski** [comment]
Seems like something we should clearly do. The only issue is the implementation. Does this approach seem good to you, @JeffBezanson?

**Comment by StefanKarpinski** [comment]
Is this ready to go, @tkf?

**Review by findmyway — APPROVED** [review]

File: stdlib/Distributed/src/remotecall.jl [hunk 1]
```diff
@@ -643,3 +643,17 @@ function getindex(r::RemoteChannel, args...)
     end
     return remotecall_fetch(getindex, r.where, r, args...)
 end
+
+function Base.iterate(c::RemoteChannel, state=nothing)
+    try
+        return (take!(c), nothing)
+    catch e
+        if isa(e, InvalidStateException) && e.state == :closed
```
Comment (tkf) [inline stdlib/Distributed/src/remotecall.jl:651]: `isa(e, InvalidStateException)` is not enough.

We'd need to handle the case when the channel is a reference

```
julia> ch = RemoteChannel(workers()[end])
RemoteChannel{Channel{Any}}(2, 1, 8)

julia> put!(ch, 1)
RemoteChannel{Channel{Any}}(2, 1, 8)

julia> close(ch)

julia> take!(ch)
1

julia> err = try
           take!(ch)
       catch err
           err
       end
RemoteException(2, CapturedException(InvalidStateException("Channel is closed.", :closed), Any[(check_channel_state at channels.jl:170 [inlined], 1), (take_buffered at channels.jl:388, 1), (take! at channels.jl:383, 1), (take! at remotecall.jl:599,
 1), (take_ref at remotecall.jl:610, 1), (#106 at process_messages.jl:278, 1), (run_work_thunk at process_messages.jl:63, 1)
, (macro expansion at process_messages.jl:278 [inlined], 1), (#105 at task.jl:411, 1)]))
```

**Comment by tkf** [comment]
#41966 is an alternative solution to this

**Comment by vtjnash** [comment]
Replaced by (included in) #41966

**Pull request closed without merging**